File: apps/tabtin-electron/src/renderer/src/components/chat/markdown/StreamingMarkdownSplitter.ts
/**
 * StreamingMarkdownSplitter — 将流式 Markdown 拆分为"稳定区"和"尾部区"。
 *
 * 稳定区：已经完成的顶层 Markdown 块（段落、标题、代码块、列表组等），
 *         内容不再变化，可以冻住渲染结果不参与后续 React reconciliation。
 * 尾部区：最后一个未完成的块，每帧轻量渲染。
 *
 * 分割策略：从末尾向前查找最后一个"双换行"（`\n\n`），该位置之前为稳定区。
 * 对代码块做特殊处理：如果稳定区末尾有未闭合的代码围栏，把整个代码块
 * 移到尾部区（避免截断代码块导致渲染错误）。
 */

export interface SplitResult {
  stable: string
  tail: string
}

const CODE_FENCE_RE = /^(`{3,}|~{3,})/

export function splitStreamingMarkdown(content: string): SplitResult {
  if (content.length < 200) {
    return { stable: '', tail: content }
  }

  const lastDoubleNewline = content.lastIndexOf('\n\n')

  if (lastDoubleNewline < 100) {
    return { stable: '', tail: content }
  }

  let splitPos = lastDoubleNewline + 2
  let candidate = content.slice(0, splitPos)

  let openFences = 0
  const lines = candidate.split('\n')
  for (const line of lines) {
    if (CODE_FENCE_RE.test(line.trimStart())) {
      openFences++
    }
  }

  if (openFences % 2 !== 0) {
    const fencePositions: number[] = []
    let pos = 0
    for (const line of lines) {
      if (CODE_FENCE_RE.test(line.trimStart())) {
        fencePositions.push(pos)
      }
      pos += line.length + 1
    }
    const lastOpenFence = fencePositions[fencePositions.length - 1]
    if (lastOpenFence > 0) {
      splitPos = lastOpenFence
      candidate = content.slice(0, splitPos)
    } else {
      return { stable: '', tail: content }
    }
  }

  return {
    stable: candidate,
    tail: content.slice(splitPos),
  }
}

File: apps/tabtin-android/app/src/main/java/com/tabtin/mobile/features/conversation/StreamingMarkdownSplitter.kt
package com.tabtin.mobile.features.conversation

/**
 * 流式 Markdown 增量切分，对齐 Electron `splitStreamingMarkdown` /
 * iOS `StreamingMarkdownSplitter`。
 *
 * - 稳定区：最后一个 `\n\n` 之前的完整顶层块，可冻住 Markdown 渲染树
 * - 尾部区：仍在增长的最后一块，用轻量 Text 上屏
 * - 未闭合代码围栏整块退回尾部（禁止对稳定区硬补 ```）
 */
internal object StreamingMarkdownSplitter {
    data class SplitResult(val stable: String, val tail: String)

    private const val MIN_SPLITTABLE_LENGTH = 200
    private const val MIN_STABLE_LENGTH = 100

    fun split(content: String): SplitResult {
        if (content.length < MIN_SPLITTABLE_LENGTH) {
            return SplitResult(stable = "", tail = content)
        }
        val lastDoubleNewline = content.lastIndexOf("\n\n")
        if (lastDoubleNewline < MIN_STABLE_LENGTH) {
            return SplitResult(stable = "", tail = content)
        }

        var splitPos = lastDoubleNewline + 2
        var candidate = content.substring(0, splitPos)
        val fencePositions = fenceLineOffsets(candidate)

        if (fencePositions.size % 2 != 0) {
            val lastOpenFence = fencePositions.lastOrNull() ?: return SplitResult("", content)
            if (lastOpenFence <= 0) {
                return SplitResult(stable = "", tail = content)
            }
            splitPos = lastOpenFence
            candidate = content.substring(0, splitPos)
        }

        return SplitResult(
            stable = candidate,
            tail = content.substring(splitPos),
        )
    }

    private fun fenceLineOffsets(text: String): List<Int> {
        val offsets = ArrayList<Int>()
        var pos = 0
        // 与 JS split('\n') 一致：保留尾部空段
        val lines = text.split("\n")
        for (line in lines) {
            if (isFenceLine(line)) {
                offsets.add(pos)
            }
            pos += line.length + 1
        }
        return offsets
    }

    private fun isFenceLine(line: String): Boolean {
        val trimmed = line.trimStart()
        if (trimmed.isEmpty()) return false
        val marker = trimmed[0]
        if (marker != '`' && marker != '~') return false
        var count = 0
        for (ch in trimmed) {
            if (ch != marker) break
            count++
        }
        return count >= 3
    }
}
